### ISSUE TITLE
Highlight `!` token as user-defined operator when it invokes a class's `operator bool`

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#1931]
In the following code:

```c++
struct S {
  int val;

  operator bool() {
    return val != 0;
  }
};

void foo(S s) {
  if (!s) {
  }
}
```

the `!s` means `!(s.operator bool())`.

I think it would be helpful if the `!` were highlighted as an overloaded operator, and navigation from there took you to the `operator bool`.

cc @ckandeler as you've worked on highlighting of overloaded operators

**Comment by ckandeler** [comment]
Are you sure? The "!" is just the built-in operator. What one really wants here is the invisible conversion op, but that seems difficult to achieve. Note that it's just as legitimate to want to get to operator bool if the expression is not negated, i.e. there's nothing syntactical to piggy-back on at all.

**Comment by HighCommander4** [comment]
You're right, it's not a perfect fit as the `!` operator isn't actually overloaded. I just thought it would be convenient to use it as a place to stick a reference to the `operator bool` in a situation like this (and you're totally right that there will be other cases like `if (s)` where there isn't a token available to host the reference).

In that sense, this is a less a semantic highlighting enhancement and more a go-to-definition enhancement (with a possible side effect of giving the `!` a semantic token).
